group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java

```diff
@@ -1932,6 +1961,435 @@ public void testPartitionAssignorExceptionOnRegularHeartbeat() {
+            new ConsumerGroupMember.Builder("group2-m1")
+                .setSubscribedTopicNames(Arrays.asList("foo", "bar", "zar"))
+                .build()));
+
+        assertEquals(mkSet(groupId1, groupId2), context.groupMetadataManager.groupsSubscribedToTopic("foo"));
+        assertEquals(mkSet(groupId1, groupId2), context.groupMetadataManager.groupsSubscribedToTopic("bar"));
+        assertEquals(mkSet(groupId2), context.groupMetadataManager.groupsSubscribedToTopic("zar"));
+
+        // M2 in group 1 subscribes to bar and zar.
+        context.replay(RecordHelpers.newMemberSubscriptionRecord(groupId1,
+            new ConsumerGroupMember.Builder("group1-m2")
+                .setSubscribedTopicNames(Arrays.asList("bar", "zar"))
+                .build()));
+
+        assertEquals(mkSet(groupId1, groupId2), context.groupMetadataManager.groupsSubscribedToTopic("foo"));
+        assertEquals(mkSet(groupId1, groupId2), context.groupMetadataManager.groupsSubscribedToTopic("bar"));
+        assertEquals(mkSet(groupId1, groupId2), context.groupMetadataManager.groupsSubscribedToTopic("zar"));
+
+        // M2 in group 2 subscribes to foo and bar.
+        context.replay(RecordHelpers.newMemberSubscriptionRecord(groupId2,
+            new ConsumerGroupMember.Builder("group2-m2")
+                .setSubscribedTopicNames(Arrays.asList("foo", "bar"))
+                .build()));
+
+        assertEquals(mkSet(groupId1, groupId2), context.groupMetadataManager.groupsSubscribedToTopic("foo"));
+        assertEquals(mkSet(groupId1, groupId2), context.groupMetadataManager.groupsSubscribedToTopic("bar"));
+        assertEquals(mkSet(groupId1, groupId2), context.groupMetadataManager.groupsSubscribedToTopic("zar"));
+
+        // M1 in group 1 is removed.
+        context.replay(RecordHelpers.newCurrentAssignmentTombstoneRecord(groupId1, "group1-m1"));
+        context.replay(RecordHelpers.newMemberSubscriptionTombstoneRecord(groupId1, "group1-m1"));
+
+        assertEquals(mkSet(groupId2), context.groupMetadataManager.groupsSubscribedToTopic("foo"));
+        assertEquals(mkSet(groupId1, groupId2), context.groupMetadataManager.groupsSubscribedToTopic("bar"));
+        assertEquals(mkSet(groupId1, groupId2), context.groupMetadataManager.groupsSubscribedToTopic("zar"));
+
+        // M1 in group 2 subscribes to nothing.
+        context.replay(RecordHelpers.newMemberSubscriptionRecord(groupId2,
+            new ConsumerGroupMember.Builder("group2-m1")
+                .setSubscribedTopicNames(Collections.emptyList())
+                .build()));
+
+        assertEquals(mkSet(groupId2), context.groupMetadataManager.groupsSubscribedToTopic("foo"));
+        assertEquals(mkSet(groupId1, groupId2), context.groupMetadataManager.groupsSubscribedToTopic("bar"));
+        assertEquals(mkSet(groupId1), context.groupMetadataManager.groupsSubscribedToTopic("zar"));
+
+        // M2 in group 2 subscribes to foo.
```

Comment:
this is effectively just removing bar from the subscription right?

Reply:
right.